// ableton-js/util/color.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Color = void 0;
/** Represents a color in Ableton */
var Color = /** @class */ (function () {
    function Color(color) {
        if (typeof color === "number") {
            this.color = color.toString(16).padStart(6, "0");
        }
        else if (color.length === 6 || color.length === 7) {
            this.color = color.replace("#", "");
        }
        else {
            throw new Error("Color " + color + " is not in a valid format");
        }
    }
    Object.defineProperty(Color.prototype, "hex", {
        get: function () {
            return "#" + this.color;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Color.prototype, "rgb", {
        get: function () {
            return {
                r: parseInt(this.color.substr(0, 2), 16),
                g: parseInt(this.color.substr(2, 2), 16),
                b: parseInt(this.color.substr(4, 2), 16),
            };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Color.prototype, "numberRepresentation", {
        get: function () {
            return parseInt(this.color, 16);
        },
        enumerable: false,
        configurable: true
    });
    Color.prototype.toString = function () {
        return this.hex;
    };
    Color.prototype.toJSON = function () {
        return this.numberRepresentation;
    };
    return Color;
}());
exports.Color = Color;
