// ableton-js/ns/clip.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Clip = exports.LaunchQuantization = exports.LaunchMode = exports.WarpMode = void 0;
var _1 = require(".");
var color_1 = require("../util/color");
var note_1 = require("../util/note");
var WarpMode;
(function (WarpMode) {
    WarpMode[WarpMode["Beats"] = 0] = "Beats";
    WarpMode[WarpMode["Tones"] = 1] = "Tones";
    WarpMode[WarpMode["Texture"] = 2] = "Texture";
    WarpMode[WarpMode["Repitch"] = 3] = "Repitch";
    WarpMode[WarpMode["Complex"] = 4] = "Complex";
    WarpMode[WarpMode["ComplexPro"] = 6] = "ComplexPro";
})(WarpMode = exports.WarpMode || (exports.WarpMode = {}));
var LaunchMode;
(function (LaunchMode) {
    LaunchMode[LaunchMode["Trigger"] = 0] = "Trigger";
    LaunchMode[LaunchMode["Gate"] = 1] = "Gate";
    LaunchMode[LaunchMode["Toggle"] = 2] = "Toggle";
    LaunchMode[LaunchMode["Repeat"] = 3] = "Repeat";
})(LaunchMode = exports.LaunchMode || (exports.LaunchMode = {}));
var LaunchQuantization;
(function (LaunchQuantization) {
    LaunchQuantization[LaunchQuantization["QGlobal"] = 0] = "QGlobal";
    LaunchQuantization[LaunchQuantization["QNone"] = 1] = "QNone";
    LaunchQuantization[LaunchQuantization["Q8Bars"] = 2] = "Q8Bars";
    LaunchQuantization[LaunchQuantization["Q4Bars"] = 3] = "Q4Bars";
    LaunchQuantization[LaunchQuantization["Q2Bars"] = 4] = "Q2Bars";
    LaunchQuantization[LaunchQuantization["QBar"] = 5] = "QBar";
    LaunchQuantization[LaunchQuantization["QHalf"] = 6] = "QHalf";
    LaunchQuantization[LaunchQuantization["QHalfTriplet"] = 7] = "QHalfTriplet";
    LaunchQuantization[LaunchQuantization["QQuarter"] = 8] = "QQuarter";
    LaunchQuantization[LaunchQuantization["QQuarterTriplet"] = 9] = "QQuarterTriplet";
    LaunchQuantization[LaunchQuantization["QEighth"] = 10] = "QEighth";
    LaunchQuantization[LaunchQuantization["QEighthTriplet"] = 11] = "QEighthTriplet";
    LaunchQuantization[LaunchQuantization["QSixteenth"] = 12] = "QSixteenth";
    LaunchQuantization[LaunchQuantization["QSixteenthTriplet"] = 13] = "QSixteenthTriplet";
    LaunchQuantization[LaunchQuantization["QThirtySecond"] = 14] = "QThirtySecond";
})(LaunchQuantization = exports.LaunchQuantization || (exports.LaunchQuantization = {}));
/**
 * This class represents an entry in Live's Session view matrix.
 */
var Clip = /** @class */ (function (_super) {
    __extends(Clip, _super);
    function Clip(ableton, raw) {
        var _this = _super.call(this, ableton, "clip", raw.id) || this;
        _this.raw = raw;
        _this.transformers = {
            color: function (c) { return new color_1.Color(c); },
            selected_notes: function (n) { return n.map(note_1.tupleToNote); },
        };
        return _this;
    }
    /**
     * Available for audio clips only.
     * Converts the given beat time to sample time.
     * Raises an error if the sample is not warped.
     */
    Clip.prototype.beatToSampleTime = function (beats) {
        return this.sendCommand("beat_to_sample_time", [beats]);
    };
    /**
     * Clears all envelopes for this clip.
     */
    Clip.prototype.clearAllEnvelopes = function () {
        return this.sendCommand("clear_all_envelopes");
    };
    /**
     * Clears the envelope of this clip's given parameter.
     */
    Clip.prototype.clearEnvelope = function (parameter) {
        return this.sendCommand("clear_envelope", { parameter: parameter });
    };
    /**
     * Creates an envelope for a given parameter and returns it.
     * This should only be used if the envelope doesn't exist.
     * Raises an error if the the envelope can't be created.
     */
    Clip.prototype.createAutomationEnvelope = function () { };
    /**
     * Crops the clip. The region that is cropped depends on whether
     * the clip is looped or not. If looped, the region outside of
     * the loop is removed. If not looped, the region outside
     * the start and end markers is removed.
     */
    Clip.prototype.crop = function () {
        return this.sendCommand("crop");
    };
    /**
     * Deselects all notes present in the clip.
     */
    Clip.prototype.deselectAllNotes = function () {
        return this.sendCommand("deselect_all_notes");
    };
    /**
     * Makes the loop twice as long and duplicates notes and envelopes.
     * Duplicates the clip start/end range if the clip is not looped.
     */
    Clip.prototype.duplicateLoop = function () {
        return this.sendCommand("duplicate_loop");
    };
    /**
     * Duplicates the notes in the specified region to the destination_time.
     * Only notes of the specified pitch are duplicated if pitch is not -1.
     * If the transposition_amount is not 0, the notes in the region will be
     * transposed by the transposition_amount of semitones.
     * Raises an error on audio clips.
     */
    Clip.prototype.duplicateRegion = function (start, length, destinationTime, pitch, transpositionAmount) {
        if (pitch === void 0) { pitch = -1; }
        if (transpositionAmount === void 0) { transpositionAmount = 0; }
        return this.sendCommand("duplicate_region", [
            start,
            length,
            destinationTime,
            pitch,
            transpositionAmount,
        ]);
    };
    /**
     * Starts playing this clip.
     */
    Clip.prototype.fire = function () {
        return this.sendCommand("fire");
    };
    /**
     * Returns all notes that match the given range.
     */
    Clip.prototype.getNotes = function (fromTime, fromPitch, timeSpan, pitchSpan) {
        return __awaiter(this, void 0, void 0, function () {
            var notes;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.sendCommand("get_notes", [
                            fromTime,
                            fromPitch,
                            timeSpan,
                            pitchSpan,
                        ])];
                    case 1:
                        notes = _a.sent();
                        return [2 /*return*/, notes.map(note_1.tupleToNote)];
                }
            });
        });
    };
    /**
     * Jump forward or backward by the specified relative amount in beats.
     * Will do nothing if the clip is not playing.
     */
    Clip.prototype.movePlayingPos = function (amount) {
        return this.sendCommand("move_playing_pos", [amount]);
    };
    /**
     * Quantizes all notes in a clip or aligns warp markers.
     */
    Clip.prototype.quantize = function (grid, amount) {
        return this.sendCommand("quantize", [grid, amount]);
    };
    /**
     * Quantizes all the notes of a given pitch.
     */
    Clip.prototype.quantizePitch = function (pitch, grid, amount) {
        return this.sendCommand("quantize_pitch", [pitch, grid, amount]);
    };
    /**
     * Deletes all notes that start in the given area.
     */
    Clip.prototype.removeNotes = function (fromTime, fromPitch, timeSpan, pitchSpan) {
        return this.sendCommand("remove_notes", [
            fromTime,
            fromPitch,
            timeSpan,
            pitchSpan,
        ]);
    };
    /**
     * Replaces selected notes with an array of new notes.
     */
    Clip.prototype.replaceSelectedNotes = function (notes) {
        return this.sendCommand("replace_selected_notes", {
            notes: notes.map(note_1.noteToTuple),
        });
    };
    /**
     * Available for audio clips only.
     * Converts the given sample time to beat time.
     * Raises an error if the sample is not warped.
     */
    Clip.prototype.sampleToBeatTime = function (sampleTime) {
        return this.sendCommand("sample_to_beat_time", [sampleTime]);
    };
    /**
     * Scrubs inside a clip.
     * `position` defines the position in beats that the scrub will start from.
     * The scrub will continue until `stop_scrub` is called.
     * Global quantization applies to the scrub's position and length.
     */
    Clip.prototype.scrub = function (position) {
        return this.sendCommand("scrub", [position]);
    };
    /**
     * Available for audio clips only.
     * Converts the given seconds to sample time.
     * Raises an error if the sample is warped.
     */
    Clip.prototype.secondsToSampleTime = function (seconds) {
        return this.sendCommand("seconds_to_sample_time", [seconds]);
    };
    /**
     * Selects all notes present in the clip.
     */
    Clip.prototype.selectAllNotes = function () {
        return this.sendCommand("select_all_notes");
    };
    /**
     * Set the clip's fire button state directly.
     * Supports all launch modes.
     */
    Clip.prototype.setFireButtonState = function (state) {
        return this.sendCommand("set_fire_button_state", [state]);
    };
    /**
     * Adds the given notes to the clip.
     */
    Clip.prototype.setNotes = function (notes) {
        return this.sendCommand("set_notes", { notes: notes.map(note_1.noteToTuple) });
    };
    /**
     * Stop playig this clip.
     */
    Clip.prototype.stop = function () {
        return this.sendCommand("stop");
    };
    /**
     * Stops the current scrub.
     */
    Clip.prototype.stopScrub = function () {
        return this.sendCommand("stop_scrub");
    };
    return Clip;
}(_1.Namespace));
exports.Clip = Clip;
