// ableton-js/ns/clip-slot.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClipSlot = exports.PlayingStatus = void 0;
var _1 = require(".");
var color_1 = require("../util/color");
var clip_1 = require("./clip");
var PlayingStatus;
(function (PlayingStatus) {
    PlayingStatus["Stopped"] = "stopped";
    PlayingStatus["Playing"] = "playing";
    PlayingStatus["Recording"] = "recording";
})(PlayingStatus = exports.PlayingStatus || (exports.PlayingStatus = {}));
/**
 * This class represents an entry in Live's Session view matrix.
 */
var ClipSlot = /** @class */ (function (_super) {
    __extends(ClipSlot, _super);
    function ClipSlot(ableton, raw) {
        var _this = _super.call(this, ableton, "clip_slot", raw.id) || this;
        _this.raw = raw;
        _this.transformers = {
            clip: function (c) { return (c ? new clip_1.Clip(ableton, c) : null); },
            color: function (c) { return new color_1.Color(c); },
        };
        return _this;
    }
    /**
     * Creates an empty clip with the given length in the slot.
     * Throws an error when called on non-empty slots or slots in non-MIDI tracks.
     */
    ClipSlot.prototype.createClip = function (length) {
        return this.sendCommand("create_clip", [length]);
    };
    /**
     * Removes the clip contained in the slot.
     * Raises an exception if the slot was empty.
     */
    ClipSlot.prototype.deleteClip = function () {
        return this.sendCommand("delete_clip");
    };
    ClipSlot.prototype.duplicateClipTo = function (slot) {
        return this.sendCommand("duplicate_clip_to", { slot_id: slot.raw.id });
    };
    /**
     * Fire a Clip if this Clipslot owns one,
     * else trigger the stop button, if we have one.
     */
    ClipSlot.prototype.fire = function () {
        return this.sendCommand("fire");
    };
    /**
     * Set the ClipSlot's fire button state directly.
     * Supports all launch modes.
     */
    ClipSlot.prototype.setFireButtonState = function (state) {
        return this.sendCommand("set_fire_button_state", [state]);
    };
    /**
     * Stop playing the contained Clip,
     * if there is a Clip and its currently playing.
     */
    ClipSlot.prototype.stop = function () {
        return this.sendCommand("stop");
    };
    return ClipSlot;
}(_1.Namespace));
exports.ClipSlot = ClipSlot;
