// ableton-js/ns/device-parameter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceParameter = exports.ParameterState = exports.AutomationState = void 0;
var _1 = require(".");
var AutomationState;
(function (AutomationState) {
    AutomationState[AutomationState["None"] = 0] = "None";
    AutomationState[AutomationState["Playing"] = 1] = "Playing";
    AutomationState[AutomationState["Overridden"] = 2] = "Overridden";
})(AutomationState = exports.AutomationState || (exports.AutomationState = {}));
var ParameterState;
(function (ParameterState) {
    ParameterState[ParameterState["Enabled"] = 0] = "Enabled";
    ParameterState[ParameterState["Disabled"] = 1] = "Disabled";
    ParameterState[ParameterState["Irrelevant"] = 2] = "Irrelevant";
})(ParameterState = exports.ParameterState || (exports.ParameterState = {}));
var DeviceParameter = /** @class */ (function (_super) {
    __extends(DeviceParameter, _super);
    function DeviceParameter(ableton, raw) {
        var _this = _super.call(this, ableton, "device-parameter", raw.id) || this;
        _this.raw = raw;
        return _this;
    }
    return DeviceParameter;
}(_1.Namespace));
exports.DeviceParameter = DeviceParameter;
