// ableton-js/ns/song.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Song = exports.RecordingQuantization = exports.Quantization = exports.TimeFormat = void 0;
var _1 = require(".");
var track_1 = require("./track");
var cue_point_1 = require("./cue-point");
var song_view_1 = require("./song-view");
var scene_1 = require("./scene");
var TimeFormat;
(function (TimeFormat) {
    TimeFormat[TimeFormat["MsTime"] = 0] = "MsTime";
    TimeFormat[TimeFormat["Smpte24"] = 1] = "Smpte24";
    TimeFormat[TimeFormat["Smpte25"] = 2] = "Smpte25";
    TimeFormat[TimeFormat["Smpte29"] = 3] = "Smpte29";
    TimeFormat[TimeFormat["Smpte30"] = 4] = "Smpte30";
    TimeFormat[TimeFormat["Smpte30Drop"] = 5] = "Smpte30Drop";
})(TimeFormat = exports.TimeFormat || (exports.TimeFormat = {}));
var Quantization;
(function (Quantization) {
    Quantization["q_2_bars"] = "q_2_bars";
    Quantization["q_4_bars"] = "q_4_bars";
    Quantization["q_8_bars"] = "q_8_bars";
    Quantization["q_bar"] = "q_bar";
    Quantization["q_eight"] = "q_eight";
    Quantization["q_eight_triplet"] = "q_eight_triplet";
    Quantization["q_half"] = "q_half";
    Quantization["q_half_triplet"] = "q_half_triplet";
    Quantization["q_no_q"] = "q_no_q";
    Quantization["q_quarter"] = "q_quarter";
    Quantization["q_quarter_triplet"] = "q_quarter_triplet";
    Quantization["q_sixtenth"] = "q_sixtenth";
    Quantization["q_sixtenth_triplet"] = "q_sixtenth_triplet";
    Quantization["q_thirtytwoth"] = "q_thirtytwoth";
})(Quantization = exports.Quantization || (exports.Quantization = {}));
var RecordingQuantization;
(function (RecordingQuantization) {
    RecordingQuantization["rec_q_eight"] = "rec_q_eight";
    RecordingQuantization["rec_q_eight_eight_triplet"] = "rec_q_eight_eight_triplet";
    RecordingQuantization["rec_q_eight_triplet"] = "rec_q_eight_triplet";
    RecordingQuantization["rec_q_no_q"] = "rec_q_no_q";
    RecordingQuantization["rec_q_quarter"] = "rec_q_quarter";
    RecordingQuantization["rec_q_sixtenth"] = "rec_q_sixtenth";
    RecordingQuantization["rec_q_sixtenth_sixtenth_triplet"] = "rec_q_sixtenth_sixtenth_triplet";
    RecordingQuantization["rec_q_sixtenth_triplet"] = "rec_q_sixtenth_triplet";
    RecordingQuantization["rec_q_thirtysecond"] = "rec_q_thirtysecond";
})(RecordingQuantization = exports.RecordingQuantization || (exports.RecordingQuantization = {}));
var Song = /** @class */ (function (_super) {
    __extends(Song, _super);
    function Song(ableton) {
        var _this = _super.call(this, ableton, "song") || this;
        _this.view = new song_view_1.SongView(_this.ableton);
        _this.transformers = {
            cue_points: function (points) { return points.map(function (c) { return new cue_point_1.CuePoint(_this.ableton, c); }); },
            master_track: function (track) { return new track_1.Track(_this.ableton, track); },
            return_tracks: function (tracks) { return tracks.map(function (t) { return new track_1.Track(_this.ableton, t); }); },
            tracks: function (tracks) { return tracks.map(function (t) { return new track_1.Track(_this.ableton, t); }); },
            visible_tracks: function (tracks) { return tracks.map(function (t) { return new track_1.Track(_this.ableton, t); }); },
            scenes: function (scenes) { return scenes.map(function (s) { return new scene_1.Scene(_this.ableton, s); }); },
        };
        return _this;
    }
    Song.prototype.beginUndoStep = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("begin_undo_step")];
            });
        });
    };
    Song.prototype.continuePlaying = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("continue_playing")];
            });
        });
    };
    Song.prototype.createAudioTrack = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("create_audio_track", [index])];
            });
        });
    };
    Song.prototype.createMidiTrack = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("create_midi_track", [index])];
            });
        });
    };
    Song.prototype.createReturnTrack = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("create_return_track", [index])];
            });
        });
    };
    Song.prototype.createScene = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("create_scene", [index])];
            });
        });
    };
    Song.prototype.deleteReturnTrack = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("delete_return_track", [index])];
            });
        });
    };
    Song.prototype.deleteScene = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("delete_scene", [index])];
            });
        });
    };
    Song.prototype.deleteTrack = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("delete_track", [index])];
            });
        });
    };
    Song.prototype.duplicateScene = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("duplicate_scene", [index])];
            });
        });
    };
    Song.prototype.duplicateTrack = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("duplicate_track", [index])];
            });
        });
    };
    Song.prototype.endUndoStep = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("end_undo_step")];
            });
        });
    };
    Song.prototype.getData = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("get_data", { key: key })];
            });
        });
    };
    Song.prototype.getCurrentSmpteSongTime = function (timeFormat) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("get_current_smpte_song_time", { timeFormat: timeFormat })];
            });
        });
    };
    Song.prototype.isCuePointSelected = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("is_cue_point_selected")];
            });
        });
    };
    Song.prototype.jumpBy = function (amount) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("jump_by", [amount])];
            });
        });
    };
    Song.prototype.jumpToNextCue = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("jump_to_next_cue")];
            });
        });
    };
    Song.prototype.jumpToPrevCue = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("jump_to_prev_cue")];
            });
        });
    };
    Song.prototype.playSelection = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("play_selection")];
            });
        });
    };
    Song.prototype.redo = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("redo")];
            });
        });
    };
    Song.prototype.scrubBy = function (amount) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("scrub_by", [amount])];
            });
        });
    };
    Song.prototype.setData = function (key, value) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("set_data", [key, value])];
            });
        });
    };
    Song.prototype.setOrDeleteCue = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("set_or_delete_cue")];
            });
        });
    };
    Song.prototype.startPlaying = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("start_playing")];
            });
        });
    };
    Song.prototype.stopAllClips = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("stop_all_clips")];
            });
        });
    };
    Song.prototype.stopPlaying = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("stop_playing")];
            });
        });
    };
    Song.prototype.tapTempo = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("tap_tempo")];
            });
        });
    };
    Song.prototype.undo = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendCommand("undo")];
            });
        });
    };
    return Song;
}(_1.Namespace));
exports.Song = Song;
