// ableton-js/ns/track.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Track = void 0;
var _1 = require(".");
var device_1 = require("./device");
var clip_slot_1 = require("./clip-slot");
var clip_1 = require("./clip");
var color_1 = require("../util/color");
var Track = /** @class */ (function (_super) {
    __extends(Track, _super);
    function Track(ableton, raw) {
        var _this = _super.call(this, ableton, "track", raw.id) || this;
        _this.raw = raw;
        _this.transformers = {
            arrangement_clips: function (clips) {
                return clips.map(function (clip) { return new clip_1.Clip(ableton, clip); });
            },
            color: function (c) { return new color_1.Color(c); },
            devices: function (devices) { return devices.map(function (d) { return new device_1.Device(ableton, d); }); },
            clip_slots: function (clip_slots) {
                return clip_slots.map(function (c) { return new clip_slot_1.ClipSlot(ableton, c); });
            },
        };
        return _this;
    }
    Track.prototype.duplicateClipToArrangement = function (clipID, time) {
        return this.sendCommand("duplicate_clip_to_arrangement", {
            clip_id: clipID,
            time: time
        });
    };
    return Track;
}(_1.Namespace));
exports.Track = Track;
